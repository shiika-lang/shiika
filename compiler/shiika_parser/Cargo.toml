[package]
name = "shiika_parser"
version = "0.1.0"
edition = "2018"

[dependencies]
shiika_core = { path = "../shiika_core" }
shiika_ast = { path = "../shiika_ast" }
thiserror = "1.0"
