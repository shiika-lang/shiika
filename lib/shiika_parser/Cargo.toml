[package]
name = "shiika_parser"
version = "0.1.0"
edition = "2021"

[dependencies]
shiika_core = { path = "../shiika_core" }
shiika_ast = { path = "../shiika_ast" }
thiserror = "1.0"
