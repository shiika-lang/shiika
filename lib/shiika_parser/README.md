# shiika_parser

This crate provides parser of Shiika programs.

The parser is hand-written (i.e. no parser generator used.)

Note: structs for ASTs (including `Token`, `Location`, `LocationSpan`, etc.) are in `lib/shiika_ast`, not here.
