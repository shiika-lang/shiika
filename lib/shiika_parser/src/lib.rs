/// Parser
///
/// Implementaion rules
/// - Call `skip_ws`/`skip_wsn` before calling other `parse_xx`

/// Create ParseError with `format!`
macro_rules! parse_error {
    ( $self:ident, $( $arg:expr ),* ) => ({
        let msg = format!( $( $arg ),* );
        $self.parseerror(&msg)
    })
}

mod ast_builder;
mod base;
mod definition_parser;
mod error;
mod expression_parser;
pub mod lexer;
mod source_file;
use crate::ast_builder::AstBuilder;
pub use crate::error::Error;
use crate::lexer::Lexer;
use crate::lexer::LexerState;
pub use crate::source_file::SourceFile;
use shiika_ast as ast;
use shiika_ast::Token;

pub struct Parser<'a> {
    pub lexer: Lexer<'a>,
    ast: AstBuilder,
    /// For debug print
    pub lv: usize,
}

impl<'a> Parser<'a> {
    pub fn new(file: &'a SourceFile) -> Parser<'a> {
        Parser {
            lexer: Lexer::new(&file.content),
            ast: AstBuilder::new(&file.path),
            lv: 0,
        }
    }

    /// Parse a method signature
    pub fn parse_signature(sig_str: &str) -> Result<ast::AstMethodSignature, Error> {
        let lexer = Lexer::new_with_state(sig_str, LexerState::MethodName);
        let mut parser = Parser {
            lexer,
            ast: AstBuilder::empty(),
            lv: 0,
        };
        let (ast_sig, _) = parser.parse_method_signature()?;
        // Check if entire string is consumed
        parser.expect_eof()?;
        Ok(ast_sig)
    }

    pub fn parse_files(files: &[SourceFile]) -> Result<ast::Program, Error> {
        let mut program = ast::Program::default();
        for file in files {
            let mut parser = Parser::new(file);
            program.append(&mut parser.parse_program()?);
        }
        Ok(program)
    }

    fn parse_program(&mut self) -> Result<ast::Program, Error> {
        self.skip_wsn()?;
        let toplevel_items = self.parse_toplevel_items()?;
        self.expect_eof()?;
        Ok(ast::Program { toplevel_items })
    }

    pub fn expect_eof(&self) -> Result<(), Error> {
        if *self.current_token() != Token::Eof {
            return Err(parse_error!(
                self,
                "unexpected token: {:?}",
                self.current_token()
            ));
        }
        Ok(())
    }

    fn parse_toplevel_items(&mut self) -> Result<Vec<ast::TopLevelItem>, Error> {
        let mut items = vec![];
        let mut base_seen = false;
        loop {
            match self.current_token() {
                Token::KwRequire => {
                    self.skip_require()?;
                }
                Token::KwBase => {
                    self.consume(Token::KwBase)?;
                    self.skip_ws()?;
                    base_seen = true;
                }
                Token::KwClass => {
                    items.push(ast::TopLevelItem::Def(
                        self.parse_class_definition(base_seen)?,
                    ));
                    base_seen = false;
                }
                Token::KwModule => {
                    items.push(ast::TopLevelItem::Def(self.parse_module_definition()?));
                }
                Token::KwEnum => {
                    items.push(ast::TopLevelItem::Def(self.parse_enum_definition()?));
                }
                Token::KwDef => {
                    return Err(parse_error!(
                        self,
                        "you cannot define toplevel method in Shiika"
                    ));
                }
                Token::Eof | Token::KwEnd => break,
                _ => {
                    let expr = self.parse_expr()?;
                    if let Some(constdef) = expr.as_const_def() {
                        items.push(ast::TopLevelItem::Def(constdef));
                    } else {
                        items.push(ast::TopLevelItem::Expr(expr));
                    }
                }
            }
            self.skip_wsn()?;
        }
        Ok(items)
    }

    /// Skip `require "foo"`
    fn skip_require(&mut self) -> Result<(), Error> {
        assert!(self.consume(Token::KwRequire)?);
        self.skip_ws()?;
        match self.current_token() {
            Token::Str(_) | Token::StrWithInterpolation { .. } => self.consume_token()?,
            _ => return Err(parse_error!(self, "unexpected argument for require")),
        };
        Ok(())
    }
}
