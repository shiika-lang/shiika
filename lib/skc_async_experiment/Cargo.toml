[package]
name = "skc_async_experiment"
version = "0.1.0"
edition = "2021"

[dependencies]
shiika_ast = { path = "../shiika_ast" }
shiika_parser = { path = "../shiika_parser" }
shiika_ffi = { path = "../shiika_ffi" }
inkwell = { git = "https://github.com/TheDan64/inkwell", features = ["llvm16-0"], rev = "4030f76" }
nom = "7.1.3"
peg = "0.8.2"
ariadne = "0.3.0"
anyhow = "1.0"
either = "1.9.0"
nom_locate = "4.2.0"
os_info = "3.7.0"
# To read exports.json5
json5 = "0.2.8"

[dev-dependencies]
insta = { version = "1.32.0", features = ["glob"] }
