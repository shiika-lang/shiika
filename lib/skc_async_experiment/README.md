# skc_async_experiment

-> doc/new_runtime.md
