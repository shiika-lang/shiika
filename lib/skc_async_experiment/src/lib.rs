mod build;
mod cli;
mod codegen;
pub mod hir;
mod hir_building;
mod hir_to_mir;
mod linker;
mod mir;
pub mod mir_lowering;
mod names;
mod package;
pub mod prelude;
pub mod run;
mod targets;
