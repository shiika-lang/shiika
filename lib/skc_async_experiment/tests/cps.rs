use anyhow::Result;
use insta::{assert_snapshot, glob};
use shiika_parser::{Parser, SourceFile};
use skc_async_experiment::{hir, hir_lowering, prelude};
use std::path::Path;

#[test]
fn test_cps_transformation() -> Result<()> {
    let base = Path::new(".").canonicalize()?;
    glob!("cps/**/*.sk", |sk_path_| {
        // Make the path relative to the project root so that the resulting .snap will be
        // identical on my machine and in the CI environment.
        let sk_path = sk_path_.strip_prefix(&base).unwrap();
        assert_snapshot!(compile(sk_path).unwrap());
    });
    Ok(())
}

fn compile(sk_path: &Path) -> Result<String> {
    let txt = std::fs::read_to_string(sk_path).unwrap();
    let src = SourceFile::new(sk_path.to_path_buf(), txt);
    let ast = Parser::parse_files(&[src])?;
    let mut hir = hir::untyped::create(&ast)?;
    hir.externs = prelude::lib_externs(Path::new("../skc_runtime/"))?
        .into_iter()
        .map(|(name, fun_ty)| hir::Extern { name, fun_ty })
        .collect();
    hir::typing::run(&mut hir)?;
    hir = hir::asyncness_check::run(hir);
    hir = hir_lowering::async_splitter::run(hir)?;
    let output = hir.to_string();
    Ok(output)
}
