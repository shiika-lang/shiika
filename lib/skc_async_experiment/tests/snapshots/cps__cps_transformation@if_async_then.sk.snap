---
source: lib/skc_async_experiment/tests/cps.rs
expression: compile(sk_path).unwrap()
input_file: lib/skc_async_experiment/tests/cps/if_async_then.sk
---
extern([+]) print (Int)->Void;
extern() sleep_sec (ChiikaEnv,Int,(ChiikaEnv,Void)->RustFuture)->RustFuture;
extern([+]) Int#+ (Int,Int)->Int;
extern([+]) Int#- (Int,Int)->Int;
extern([+]) Int#* (Int,Int)->Int;
extern([+]) Int#% (Int,Int)->Int;
extern([+]) Int#and (Int,Int)->Int;
extern([+]) Int#or (Int,Int)->Int;
extern([+]) Int#xor (Int,Int)->Int;
extern([+]) Int#lshift (Int,Int)->Int;
extern([+]) Int#rshift (Int,Int)->Int;
extern([+]) Int#< (Int,Int)->Bool;
extern([+]) Int#<= (Int,Int)->Bool;
extern([+]) Int#> (Int,Int)->Bool;
extern([+]) Int#>= (Int,Int)->Bool;
extern([+]) Int#== (Int,Int)->Bool;
fun run(ChiikaEnv $env, (ChiikaEnv,Int)->RustFuture $cont) -> RustFuture {
  chiika_env_push_frame(%arg0, 1)  #-> Void
  chiika_env_set(%arg0, 0, (%arg1 as Any), 6)  #-> Void
  if true
    return run't[*](%arg0) # RustFuture
  else
    return run'f[*](%arg0) # RustFuture
  end  #-> Never
}
fun run't(ChiikaEnv $env) -> RustFuture {
  return sleep_sec[*](%arg0, 1, run_2) # RustFuture  #-> Never
}
fun run_2(ChiikaEnv $env, Void $async_result) -> RustFuture {
  %arg1  #-> Void
  return run'e(%arg0, 1) # RustFuture  #-> Never
}
fun run'f(ChiikaEnv $env) -> RustFuture {
  return run'e(%arg0, 2) # RustFuture  #-> Never
}
fun run'e(ChiikaEnv $env, Int $ifResult) -> RustFuture {
  %arg1  #-> Int
  alloc $0  #-> Void
  $0 = 3  #-> Void
  return (chiika_env_pop_frame(%arg0, 1) as (ChiikaEnv,Int)->RustFuture)(%arg0, $0) # RustFuture  #-> Never
}
